ConsoleLogger for more fully featured log printing

c42f opened the pull request:
Here I've taken `SimpleLogger`, copied it into the logging module, and started on some formatting features which will make it a lot nicer to use.  I've also made it the default instead of `SimpleLogger` which we might consider making even simpler.

Primarily, the printing of key value pairs has been beefed up so that large data structures are summarized with the `:limit=>true` IOContext setting (may be disabled with a setting), and exceptions are now printed using `showerror` including a backtrace if it's supplied or inferred from the contents of `catch_backtrace()`.

Formatting of exceptions:
![exception_formatting](https://user-images.githubusercontent.com/601473/34521922-f890f336-f0db-11e7-9d30-79e99cb639f0.png)

Show limiting:
![show_limit](https://user-images.githubusercontent.com/601473/34521993-524f5de0-f0dc-11e7-8928-24a94c988d64.png)

I've also incorporated the right justification of metadata to get this out of the way, and removed the `Info` and `Debug` prefix by default, but with a setting to control this.  The current default style:
![default_style](https://user-images.githubusercontent.com/601473/34521835-9c4d2c02-f0db-11e7-9f79-8011472e0f68.png)

@fredrikekre

c42f committed:
ConsoleLogger for readable interactive logging

Start of a ConsoleLogger optimized for viewing log records in the REPL.
This is the SimpleLogger code copied and modified to:

* Provide configurable metadata formatting
* Right justify some of the metadata
* Use show() and showerror() for formatting values in ConsoleLogger key
  value pairs.
* Set the :limit IOContext key to sensibly show approximations of large
  data structures

Also make this the default logger type so that users can get interface
improvements by upgrading the stdlib Logging library rather than being
bound to Base.

fredrikekre commented:
Great! One thing that I wanted to do also is to remove the "exception = " prefix, which IMO looks a bit funny. I think it would look nicer with something like the following:
```julia
julia> @info "An error while creating the log event" sqrt(-1)
┌ Error: Exception while generating log record in module Main at REPL[1]:1
│   DomainError with -1.0:
    sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
└ @ Main REPL[1]:1
```

KristofferC commented:
Perhaps turn off the line info by default?

Also, why are we printing `Error: .... Error @`? Seems a waste to print `Error` twice.

StefanKarpinski commented:
I'd prefer to keep the `Info` and `Debug` prefixes for clarity. Otherwise looks lovely!

yurivish commented:
A couple of observations:

<img width="835" alt="screenshot 2018-01-03 14 55 23" src="https://user-images.githubusercontent.com/114388/34537371-2b4c96ae-f096-11e7-98b7-77921eb2cd95.png">

During interactive development one purpose of a traceback is to focus the user's attention on what went wrong and where it did so. From that perspective, the first useful piece of information above is in the row labeled `[2]` -- once you've seen that and correlated it back to the source line, you can start to understand the rest of the traceback.

In that light, I'm wondering how difficult it would be to emphasize the paths or lines indicating "user code", so they can draw your eye first and give you the context you need to interpret the rest of the stack trace? 

There are a number of circumstances where the distinction between "user" and "non-user" code is less clear, such as when doing package development, or testing at the REPL code written elsewhere, but if there are simple rules that would work the majority of the time... Even just highlighting the "topmost" (non-REPL-internal) user-written code would probably notably improve the debugging experience and save people a lot of time.

Also, the white `)` seems superfluous to me -- it catches the eye and doesn't supply any new information. Maybe just highlight the function name instead?

alyst commented on stdlib/Logging/src/ConsoleLogger.jl:
```diff
@@ -0,0 +1,112 @@
+    dsize = displaysize(logger.stream)
+    width = dsize[2]
+    locationstr = string(levelstr, " @ ", _module, " ", basename(filepath), ":", line)
+    singlelinewidth = 2 + length(msglines[1]) +
```
Is it guaranteed that `msglines` has at least one element? (`split("")` gives an empty array in 0.6)

alyst commented on stdlib/Logging/src/ConsoleLogger.jl:
```diff
@@ -0,0 +1,112 @@
+    singlelinewidth = 2 + length(msglines[1]) +
+                      (prefixwithlevel ? length(levelstr) + 2 : 0) +
+                      length(locationstr)
+    if length(msglines) + length(kwargs) == 1 && singlelinewidth <= width
```
If `length(msglines) == 0` and `length(kwargs) == 1`, kwarg would not be printed.

c42f replied:
True, this is also a bug in `SimpleLogger`.

c42f replied:
Actually no - `split("", '\n')` gives an array of length 1, so this is actually fine.  I'll add a corresponding test case.

alyst commented on stdlib/Logging/src/ConsoleLogger.jl:
```diff
@@ -0,0 +1,112 @@
+                end
+            end
+        end
+        print_with_color(color, iob, "└ ", bold=true)
```
For consistency with single line mode, `locationstr` could be printed at the last message/kwarg line if it fits.

c42f replied:
That would be nice. It does complicate the implementation significantly.

c42f replied:
I think I'll punt on this for now.  Getting right justification correct in general seems to involve removing ANSI escape sequences from the formatted values to determine the text width of the line prefix. It really seems somewhat messy.

c42f commented:
> One thing that I wanted to do also is to remove the "exception = " prefix

Yes there's a bit of tension here between having general key value pairs and the specific, very useful case of transporting an exception as context for a message. I'm not sure what to do about that; it seems a little less than ideal to give no indication of the key.  But it also seems redundant and noisy to have the `exception =` prefix in there.  It would be good to encourage people to use `exception` as the standard key for passing exceptions for consistency, so having it do something *useful* and visible in the printing would be nice ;-)

KristofferC commented:
The same backtrace filtering as the REPL does should likely be done here as well.

fredrikekre commented:
> it seems a little less than ideal to give no indication of the key

But in the case failure when generating the message the key `exception` is not something the user has defined, its just an internal thing, so thats why I don't like the printing of it, it looks like I had a variable called `exception` in my code.

KristofferC commented:
Piggybacking on the `key = value` printing, when no such thing existed in the user code indeed seems confusing.

It is a big difference between an exception getting thrown when creating the logging message and just capturing an exception with the `key = value` syntax.

c42f commented:
> It is a big difference between an exception getting thrown when creating the logging message and just capturing an exception with the key = value syntax.

I think we should fix this by setting the `group` metadata entry to `:logging_failure` or some such key.  This way messages resulting from failed logging will be directly and simply distinguishable by checking the group. For example, it will be straightforward to filter and find all instances of failed logging, straightforward to display them in a unique way, etc.

But I think worrying about this is a bit of a distraction: these types of messages should not be something you'll see often, and I'd like to focus this PR on getting the formatting as readable as possible in the common case.  For that, I think markdown based formatting of the log message is probably the next thing to tackle (in addition to cleaning up various issues people have pointed out above).

c42f commented:
> it looks like I had a variable called exception in my code

The underlying issue is there's currently no distinction between keys which have special meaning to the system (currently `maxlog`, `exception`, and probably `progress` in the near future) from any other variable the user may attach.  For now I thought that imposing the least structure was best, but there's a range of syntactic options or naming conventions we could adopt. Simplest would probably be to prefix all "special" log keys with an underscore. So far I just punted on solving this problem.

c42f committed:
Set the group of failed log events to logevent_error

This is to allow clean filtering of log event failures when these are
automatically caught by the system.

c42f commented:
Ok, I think I've got a much cleaner implementation of metadata printing: there's now a `meta_formatter` which can be passed on construction of `ConsoleLogger` which returns a `prefix` and `suffix` of metadata to be added to the message.  By default the prefix is `$level:` for all levels, and the suffix is `@ $module $file:$line`.  But this is now completely customizable which I think also addresses #25106 in a reasonably nice way.

Latest default format:

![consolelogger](https://user-images.githubusercontent.com/601473/34614619-ce98df82-f27d-11e7-833b-cae4ea4a1438.png)

Regarding the reporting of exceptions generated within the system, I've modified the group to `:logevent_error` for those cases which should allow them to be cleanly distinguished by any installed logger.

I think this is probably useful enough to merge now. Further changes like markdown handling and more sophisticated key value pair printing can come later.

KristofferC commented:
Looks very nice!

c42f commented:
CI failures look unrelated - freebsd failed when cloning a git repo, win32 failed when building deps.

iblis17 commented:
well, all CI failed on master:
(https://freebsdci.julialang.org/#/builders/1/builds/6159)

c42f commented:
Ok. What.  That's extremely bad, but also confusing given that CI succeeded on Travis, circleci, appveyor (win64). I'm looking into it immediately.

c42f commented:
Ah, I'm not going crazy.  This was a clash with changes introduced after the tests were run in https://github.com/JuliaLang/julia/pull/25289

ararslan commented:
Why was right justification included? It looks messy and it's hard to visually match info to line numbers when the terminal is set to wide.

fredrikekre commented:
![screenshot from 2018-01-07 23-17-02](https://user-images.githubusercontent.com/11698744/34654819-01d25a52-f401-11e7-82ac-c54fc5b66394.png)

c42f commented:
> It looks messy

So does the option of having the metadata mixed with the messages. If you want to see the message 90% of the time and the metadata 10% of the time it's better to put it out of the way where it's not distracting.

We can add an option for the maximum right justification of the metadata - set it to 80 columns or something by default.

ararslan commented:
Can we just make right justification opt-in?

c42f commented:
The design goal I'm trying to follow for `ConsoleLogger` is to make the default printing a really good debugging tool for the average user and package author at the REPL. So whatever it takes to make this happen is what we should do IMO.

My guess is the average user wants to see output more like `println()` (just their message, no decoration). From this point of view we should just remove all the metadata. But on relatively rare occasions it's really helpful so the thought was to put it away somewhere nonintrusive. Unfortunately there's not a lot of options for achieving this in a terminal.

JeffBezanson commented on stdlib/Logging/src/ConsoleLogger.jl:
```diff
@@ -0,0 +1,119 @@
+    ex,bt = e
+    showerror(io, ex, bt; backtrace = bt!=nothing)
+end
+showvalue(io, ex::Exception) = showvalue(io, (ex,catch_backtrace()))
```
I don't think we should do this --- there's no way to know whether the result of `catch_backtrace` corresponds to the given exception. This mechanism is already pretty fragile and can cause quite a bit of confusion, e.g. #19979

c42f replied:
One alternative is to have people pass it explicitly everywhere:

```julia
try
    # do stuff
catch ex
    @error "Extra information about something bad" exception=ex,catch_backtrace()
end
```

This might be an acceptable syntax burden?

c42f replied:
On further thought, to be completely safe from the problems of #19979 the pattern would have to look like

```julia
try
    # do stuff
catch ex
    bt = catch_backtrace()
    @error "Extra information about something bad" exception=ex,bt
end
```

Hmm. This breaks the general principle that log events should be lazily generated.

JeffBezanson commented:
Overall nice improvement! :+1:

c42f commented:
Aligned but not completely right justified looks like it might be a good compromise.  The following is metadata aligned to the 100th column, right justified:

![aligned_metadata](https://user-images.githubusercontent.com/601473/34656308-78bae4ba-f463-11e7-915d-563d45a5a475.png)

Sacha0 commented:
As a counterpoint, right justification strikes me as cleaner. (Also I tend to vertically split terminal windows on wide screens, making the separation a non-issue.) Best!

c42f commented:
@Sacha0 - interesting, this is how I use my terminals as well.

Another option (suggested on slack) is to use something other than spaces as padding.  Here's two examples with either full width or aligned to 100 columns.

![nonspace padding](https://user-images.githubusercontent.com/601473/34656860-69beba88-f46b-11e7-912d-653b1dc51279.png)

alyst commented:
Some editors/terminals highlight the line under cursor. `konsole` implements triple click to select the whole line, gnome terminal has the similar thing.
There's also good-old scrolling to the top/bottom edge of the screen.  
So right-justification is not fatal even for the wide screens.

> Another option (suggested on slack) is to use something other than spaces as padding.

Solid lines look a little bit too intrusive.
Maybe something like `.   .   .   .` as in the old movies? :)
Although I'm not sure it should be the default option.

c42f commented:
> Maybe something like . . . . as in the old movies?

I tried a lot of different variants of this with various characters.  It's a little more light weight, but not substantially different.

Sacha0 commented:
I find the horizontal lines quite distracting. Best!

StefanKarpinski commented:
I really think the old behavior of printing the `@ Module file:line` info right after the message in single-line and right below the message in multiline messages was less problematic. Perhaps that should just be the default be there can be knobs to configure this as people prefer?

StefanKarpinski commented:
Also, I would add that reviewing these changes has been made rather difficult by the fact that there are major functional changes mixed in with aesthetic changes. I'm all on board with adding the ConsoleLogger here but aspect of the previous incarnation of the output format were preferable.

c42f commented:
Sorry about that, this could have been a bit better factored in a couple of places. The original intention was that this be entirely aesthetic changes. Jeff's `catch_backtrace()` objection is completely reasonable but more a flaw in the original logging PR: it was the documented intent of the `exception` keyword.  Here I thought I was just making use of that existing assumption.
